packages/stripe/src/lib/stripe.module.ts

```diff
@@ -39,34 +39,49 @@ const makeController = (prefix: string, route: string) => {
   return Controller;
 };
 
-export const stripeModule = new AppModule(
+export class StripeModule extends createModule(
   {
     config: stripeModuleConfig,
-    providers: [
-      {
-        provide: 'Stripe',
-        deps: [stripeModuleConfig.all()],
-        useFactory: (config: typeof stripeModuleConfig.type) => {
-          const { apiKey } = config;
-          return new Stripe(apiKey, {
-            typescript: true,
-            apiVersion: '2020-08-27',
-          });
-        },
-      },
-    ],
-    exports: ['Stripe'],
+    exports: [Stripe],
   },
   'stripe'
-).setup((stripeModule, config) => {
-  if (config.webhookConfig) {
-    const webhookConfig = config.webhookConfig!;
-    stripeModule.addProvider(StripeWebhookService);
-    stripeModule.addController(
-      makeController(
-        webhookConfig.controllerRoutePrefix!,
-        webhookConfig.controllerRoute!
-      )
-    );
+) {
+
+  process() {
+    this.addProvider({
+      provide: Stripe,
+      useFactory: () => {
+        return new Stripe(this.config.apiKey, {
+          typescript: true,
+          apiVersion: '2020-08-27',
+        });
+      },
+    });
+
+    if (this.config.webhookConfig) {
+      this.addProvider(StripeWebhookService);
+      this.addController(
+        makeController(
+          this.config.webhookConfig.controllerRoutePrefix,
+          this.config.webhookConfig.controllerRoute,
+          this.config.webhookConfig.secret
+        )
+      );
+    }
+  }
+
+  processController(module: AppModule<any>, controller: ClassType) {
+    if (!this.config.webhookConfig) return;
+
+    const configs = stripeHandlerMap.get(controller);
+    if (!configs) return;
+
+    if (!module.isProvided(controller)) module.addProvider(controller);
```

Comment:
I saw similar code in the Framework when reviewing some of the changes. What are the implications of receiving a controller here that isn't provided by the module? 

Do you think that this should just be default behavior inside the framework so that module authors aren't required to do this?

Reply:
I knew this question would arise. :)
Only the consumer of the controller knows how it will be handled, which means it knows how to intantiate. For example, HTTP controllers will always be added to the 'http' scope, so it is required to call "addProvider({provide: controller, scope: 'http'})". In your case you don't use a scope, so you just add it. In any case, the user is able to override this behavior by manually declaring a provider for that controller. The framework doesn't know in advance in which scope it should put the controller, that's why it can't put it in providers automatically and why the controller consumer is responsible for that.

Reply:
Thanks for the clarification! That makes sense